### PR TITLE
Api discovery: Detect string formats

### PR DESCRIPTION
**Pull request by timokoessler**
This PR adds the data format of strings to the reported api data schema.

## Supported formats

- date: `2024-01-01`
- date-time: e.g. `1937-01-01T12:00:27.87+00:20`
- email
- uri
- uuid
- ipv4
- ipv6

## Links
https://swagger.io/docs/specification/v3_0/data-models/data-types/#strings

### CODE REVIEW

File: library/agent/api-discovery/getStringFormat.ts [hunk 1]
```diff
@@ -0,0 +1,88 @@
+import { isIPv4, isIPv6 } from "net";
+import isDateString from "./helpers/isDateString";
+import isDateTimeString from "./helpers/isDateTimeString";
+import isUUIDString from "./helpers/isUUIDString";
+
+/**
+ * https://swagger.io/docs/specification/v3_0/data-models/data-types/#strings
+ */
+export type StringFormat =
```
Comment (hansott) [inline library/agent/api-discovery/getStringFormat.ts:9]:
```suggestion
type StringFormat =
```

File: library/agent/api-discovery/getStringFormat.ts [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+}
+
+/**
+ * Check for indication characters in a string
+ * This is used to improve performance and is more efficient than multiple .includes() calls
+ */
+function checkForIndicationChars(str: string): Set<string> {
+  const foundChars = new Set<string>();
+
+  for (const char of str) {
```
Comment (hansott) [inline library/agent/api-discovery/getStringFormat.ts:81]: we iterate the str manually, maybe just use `include(":")` etc?

File: library/agent/api-discovery/getStringFormat.ts (hunk 1 above)
Comment (timokoessler) [reply library/agent/api-discovery/getStringFormat.ts:11]: Imported from other source file.

File: library/agent/api-discovery/helpers/isDateTimeString.ts [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+      offsetHour < 0 ||
+      offsetHour > 23 ||
+      offsetMinute < 0 ||
+      offsetMinute > 59
+    ) {
+      return false;
+    }
+  }
+
+  return true;
```
[the remark this answers is not in the document]
Comment (timokoessler) [reply library/agent/api-discovery/helpers/isDateTimeString.ts:53]: `new Date("1")` would also work

**Review by hansott — APPROVED** [review]

**Pull request merged**